lib/spack/spack/cmd/stack.py

```diff
@@ -0,0 +1,120 @@
+##############################################################################
+# Copyright (c) 2013-2017, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/spack/spack
+# Please also see the NOTICE and LICENSE files for our notice and the LGPL.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License (as
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+""" Stack spack installations - avoiding double-builds
+
+Stack this spack repository on top of an external read-only spack installation
+that already has several packages pre-built.
+
+The installed packages in the remote spack-installation are symlinked into the
+directory_layout of the current repository. Hardlinks make little to no sense
+since the original files need to remain in place because dependencies are found
+via RPATH (which are the old paths in the remote repository).
+"""
+
+from __future__ import print_function
+
+__author__ = "Oliver Breitwieser (UHEI)"
+
+import os
+import os.path as osp
+import spack
+import spack.config
+import spack.directory_layout
+import spack.store
+from spack.filesystem_view import filter_exclude
+from spack.util.path import canonicalize_path
+import llnl.util.tty as tty
+import llnl.util.filesystem as fs
+
+description = "integrate packages of remote spack repositories into this one"
+section = "admin"
+level = "long"  # TODO: re-check what 'level' is supposed to mean
+
+
+def setup_parser(sp):
```

Comment:
You really do like your 2 letter variable names^^

I feel like there should be a "check remote" command, it should essentially just check the existence of the original file for each symlink. A poor-man implementation of this might be a simple: Remove all links and perform again. There should also probably be a section on "what probably happened if stuff randomly breaks" in the help section

Reply:
Hmm.. in what way would it be different from just re-stacking? All symlinks with the same hash will be (deleted and) re-created.

I have yet to check what happends during `reindex`ing the repository with dangling symlinks.

Maybe it would be better to add a "remove all dangling symlinks"-phase to `reindex` instead?